### ISSUE TITLE
Critical security vulnerabilities in vm2

### ISSUE THREAD

**Issue by josephjclark** [OpenFn/core#47]
This issue is to track the new security vulnerabilities reported in vm2. The details we have right now are vague but it's something like:

1) Inject code into a promise via the `@@species` decorator: https://github.com/patriksimek/vm2/security/advisories/GHSA-cchq-frgv-rjh5 (More on this hack here: https://nvd.nist.gov/vuln/detail/CVE-2023-37466)
2) Inject code by the built-in node inspector https://github.com/patriksimek/vm2/security/advisories/GHSA-g644-9gfx-q4q4

I'll explore this a bit further but my feeling on the fixes is something like:

* Fail to compile any job with @@ annotations (no-one needs that stuff, let's just blanket ban it). If this is too broad we can just fail @@species annotations.
* Restrict access to the inspect module in the vm2 sandbox, if we haven't already. There's no need for anyone to try and use it.

**Comment by josephjclark** [comment]
The inspect issue is something to do with [this](https://nodejs.org/api/util.html#custom-inspection-functions-on-objects) (I barely understand this code example)

If you declare a class, and the class overrides util.inspect.custom, you can define a method to run free code. Somehow. Presumably that methods runs outside the sandbox (or at least in the global scope).
```
class Box {
  [util.inspect.custom](depth, options, inspect) {
     // run your evil code here
  }
   
}
```

`util.inspect.custom` is a symbol which must be required from `node:util`. Require is not exposed to the job through the existing sandbox. Therefore I don't think job code can exploit this.

**Comment by josephjclark** [comment]
Sadly failing to compile `@@Species` references isn't going to work. `@@species` isn't an annotation, and references to it look like this:

```
class Array {
  static get [Symbol.species]() {
    return this;
  }
}
```

We can't refuse to compile just because we see `species`, and it's easy to workaround `Symbol.species` references.

Banning classes seems drastic but might be reasonable. We just refuse to compile if we see a class declaration (or a class with a static property?). I don't imagine any of our job code uses classes and it's hard to imagine a reason why it would need to do so. It is relatively easy to say "sorry, we don't support Classes".

**Comment by taylordowns2000** [comment]
@josephjclark , please do
1. finish up the "you can't require" unit test: https://github.com/OpenFn/core/security/dependabot/15
2. ban `class` during compilation

**Comment by taylordowns2000** [comment]
closed in version `1.5.0`
